src/action/linux/provision_selinux.rs

```diff
@@ -0,0 +1,125 @@
+use std::path::{Path, PathBuf};
+
+use tokio::fs::{create_dir_all, remove_file};
+use tokio::process::Command;
+use tracing::{span, Span};
+
+use crate::action::{ActionError, ActionErrorKind, ActionTag};
+use crate::execute_command;
+
+use crate::action::{Action, ActionDescription, StatefulAction};
+
+const SE_LINUX_POLICY_PP_CONTENT: &[u8] = include_bytes!("selinux/nix.pp");
+
+/**
+Delete an operating system level user
+*/
+#[derive(Debug, serde::Deserialize, serde::Serialize, Clone)]
+pub struct ProvisionSelinux {
+    policy_path: PathBuf,
+}
+
+impl ProvisionSelinux {
+    #[tracing::instrument(level = "debug", skip_all)]
+    pub async fn plan(policy_path: PathBuf) -> Result<StatefulAction<Self>, ActionError> {
+        let this = Self { policy_path };
+
+        // Note: `restorecon` requires us to not just skip this, even if everything is in place.
+
+        Ok(StatefulAction::uncompleted(this))
+    }
+}
+
+#[async_trait::async_trait]
+#[typetag::serde(name = "provision_selinux")]
+impl Action for ProvisionSelinux {
+    fn action_tag() -> ActionTag {
+        ActionTag("provision_selinux")
+    }
+    fn tracing_synopsis(&self) -> String {
+        format!("Install an SELinux Policy for Nix")
+    }
+
+    fn tracing_span(&self) -> Span {
+        span!(
+            tracing::Level::DEBUG,
+            "provision_selinux",
+            policy_path = %self.policy_path.display()
+        )
+    }
+
+    fn execute_description(&self) -> Vec<ActionDescription> {
+        vec![ActionDescription::new(
+            self.tracing_synopsis(),
+            vec![format!(
+                "On SELinux systems (such as Fedora) a policy for Nix needs to be configured for correct operation."
+            )],
+        )]
+    }
+
+    #[tracing::instrument(level = "debug", skip_all)]
+    async fn execute(&mut self) -> Result<(), ActionError> {
+        if self.policy_path.exists() {
+            // Rebuild it.
+            remove_existing_policy(&self.policy_path)
+                .await
+                .map_err(Self::error)?;
+        }
+
+        if let Some(parent) = self.policy_path.parent() {
+            create_dir_all(&parent)
+                .await
+                .map_err(|e| ActionErrorKind::CreateDirectory(parent.into(), e))
+                .map_err(Self::error)?;
+        }
+
+        tokio::fs::write(&self.policy_path, SE_LINUX_POLICY_PP_CONTENT)
+            .await
+            .map_err(|e| ActionErrorKind::Write(self.policy_path.clone(), e))
+            .map_err(Self::error)?;
+
+        execute_command(Command::new("semodule").arg("-i").arg(&self.policy_path))
```

Comment:
Nit:

Do these SELinux binaries provide long flags for these options? It would be nice to see them instead, if possible, rather than the opaque `-i`, `-r`, etc.

Reply:
You're right!